File: secd/secd.c
#include <stddef.h>
#include <stdlib.h>
#include <stdio.h>

enum Op
{
    OP_NIL = 0, // Empty list
    OP_LD,      // Push value from environment to stack
    OP_LDC,     // Push constant to stack
    OP_LDF,     // Push combinator to stack

    OP_AP,      // Apply function
    OP_RTN,     // Push result on restored stack

    OP_DUM,     // Dummy value
    OP_RAP,     // Like AP but supports recursive function calls

    OP_SEL,     // Conditional branching
    OP_JOIN,    // Jump to label

    OP_CAR,     // Pops pair and pushes first element
    OP_CDR,     // Pops pair and pushes second element
    OP_ATOM,    // Push true if value is atomic
    OP_CONS,    // Pop two values from stack and push a pair

    OP_ADD,
    OP_EQ,

    OP_STOP     // End execution
};

typedef struct LL {
    void* data;
    struct LL* next;
} LL;

typedef struct {
    char *name;
    void* value;
} Map;

typedef struct {
    enum Op op;
    void* arg1;
    void* arg2;
} Instr;

typedef struct {
    size_t capacity;
    size_t index;
    Map *data[1024]; // 1024 is here directly link with the value passed in `new_state`
} Hashtbl;

typedef struct {
    size_t len;
    int index;
    Instr **lst;
} Array;

typedef struct State {
    LL *stack;
    Hashtbl *env;
    Array *input;
    struct State *dump;
} State;

typedef struct {
    char *name;
    void *body;
    Hashtbl *env;
} Closure;

Hashtbl*
new_hashtbl(int capacity)
{
    Hashtbl *hashtbl = malloc(sizeof(Hashtbl));
    if (!hashtbl) {
        perror("Failed to allocate hastbl");
        return NULL;
    }
    hashtbl->capacity = capacity;
    hashtbl->index = 0;
    return hashtbl;
}

State*
new_state(Instr **lst,
          size_t len)
{
    State *state = malloc(sizeof(State));
    if (!state) {
        perror("Failed to allocate State");
        return NULL;
    }
    state->stack = NULL;
    state->env = new_hashtbl(1024);
    Array *input = malloc(sizeof(Array));
    if (!input) {
        perror("Failed to allocate Array");
        free(state);
        return NULL;
    }
    input->len = len;
    input->index = 0;
    input->lst = lst;
    state->input = input;
    state->dump = NULL;
    return state;
}

Instr*
next(Array *arr)
{
    if (arr && arr->index < arr->len)
        return arr->lst[arr->index++];
    return NULL;
}

void
push_stack(State* state, void *value)
{
    LL* new_node = malloc(sizeof(LL));
    if (!new_node) {
        perror("Failed to allocate LL");
        return;
    }
    new_node->data = value;
    new_node->next = state->stack;
    state->stack = new_node;
}

Map*
new_map(char *name, void *value)
{
    Map *map = malloc(sizeof(Map));
    if (!map) {
        perror("Failed to allocate Map");
        return NULL;
    }
    map->name = name;
    map->value = value;
    return map;
}

// TODO: Add a hashtbl.(c/h) with utils functions
void
add_entry(State* state,
          char* name,
          void* value)
{
    if (state->env && state->env->index < state->env->capacity) {
        state->env->data[state->env->index] = new_map(name, value);
        state->env->index++;
    }
    fprintf(stderr, "Failed to add entry");
}

/*
 * LDF 	s e (LDF f.c) d            ->  ((f.e).s) e c d
 * AP  	((f.e') v.s) e (AP.c) d    ->  NIL (v.e') f (s e c.d)
 * RTN	(x.z) e' (RTN.q) (s e c.d) ->  (x.s) e c d
 */
void
vm(State* state)
{
    while (state->input->lst[state->input->index] != (void *)OP_STOP) {
        Instr* instr = next(state->input);

        // SECD abstract machine
        if (instr == (void*)OP_LD) {
            push_stack(state, instr->arg1);
        } else if (instr == (void*)OP_LDC) {
            push_stack(state, instr->arg1);
        } else if (instr == (void*)OP_LDF) {
            Closure *closure = malloc(sizeof(Closure));
            closure->name = instr->arg1;
            closure->body = instr->arg2;
            closure->env = state->env;
            push_stack(state, closure);
        } else if (instr == (void*)OP_AP) {

        }
    }
}

Instr test1_0 = {OP_LDC, (void*)1, NULL };
Instr test1_1 = {OP_LDC, (void*)1, NULL };
Instr test1_2 = {OP_ADD, NULL, NULL };
Instr *test1[] = {&test1_0, &test1_1, &test1_2};

int
main(void)
{
    State* state = new_state(test1, sizeof(test1)/sizeof(test1[0]));
    return 0;
}

File: secd/secd.hs
data Value
  = VAL_NIL
  | VAL_NUM Integer
  | VAL_BOOL Bool
  | VAL_PAIR Value Value
  | CLOSURE String [Instruction] Map
  deriving (Show, Eq)

data Instruction
  = OP_NIL
  | OP_LD String
  | OP_LDC Value
  | OP_LAM String [Instruction]
  | OP_LAB String [Instruction]
  | OP_REC String [Instruction]
  | OP_AP
  | OP_RET
  | OP_SEL [Instruction] [Instruction]
  | OP_CAR
  | OP_CDR
  | OP_CONS
  | OP_ATOM
  | OP_ADD
  | OP_MUL
  | OP_EQ
  | OP_STOP
  deriving (Show, Eq)

type Program = [Instruction]

type Map = [(String, Value)]

data SECD
  = SECD
      { stack :: [Value],
        env :: Map,
        code :: Program,
        dump :: SECD
      }
  | Empty
  deriving (Show, Eq)

proc :: SECD -> SECD
proc secd@SECD {stack, env, code = x : xs, dump} = case x of
  OP_NIL -> secd {stack = VAL_NIL : stack, code = xs}
  OP_LD name -> case lookup name env of
    Just val -> secd {stack = val : stack, code = xs}
    Nothing -> error $ "Undefined variable: " ++ name
  OP_LDC c -> secd {stack = c : stack, code = xs}
  OP_LAM name body ->
    let closure = CLOSURE name body env
     in secd {stack = closure : stack, code = xs}
  OP_AP -> case stack of
    (fun : arg : s') -> case fun of
      CLOSURE name body env' ->
        secd
          { stack = [],
            env = (name, arg) : env',
            code = body,
            dump = secd {stack = s', code = xs}
          }
      _ -> error "Invalid function application : Not a closure"
    _ -> error "Invalid stack for OP_AP"
  OP_RET -> case dump of
    Empty -> error "Empty dump"
    SECD {stack = s', env = e', code = cs', dump = d'} ->
      case stack of
        (res : _) -> SECD {stack = res : s', env = e', code = cs', dump = d'}
        [] -> error "Empty stack for OP_RET"
  OP_LAB name body ->
    let closure = CLOSURE name body ((name, closure) : env)
    in secd {env = (name, closure) : env, code = xs}
  OP_REC name body ->
    let recClosure = CLOSURE name body ((name, recClosure) : env)
    in secd {stack = recClosure : stack, code = xs}
  OP_SEL if_case else_case ->
    let (pred : r) = stack
     in case pred of
          VAL_BOOL bool ->
            let new_code = (if bool then if_case else else_case) ++ xs
             in secd {stack = r, code = new_code}
          _ -> error "OP_SEL expected a boolean"
  OP_CONS -> case stack of
    (x : y : rest) ->
      let pair = VAL_PAIR x y
       in secd {stack = pair : rest, code = xs}
    _ -> error "OP_CONS expected at least two arguments"
  OP_CAR -> case stack of
    (h : t) -> case h of
      VAL_PAIR car _ -> secd {stack = car : t, code = xs}
      _ -> error "OP_CAR exepected PAIR"
    _ -> error "OP_CAR expected one argument"
  OP_CDR -> case stack of
    (h : t) -> case h of
      VAL_PAIR _ cdr -> secd {stack = cdr : t, code = xs}
      _ -> error "OP_CDR expected PAIR"
    _ -> error "OP_CDR expected one argument"
  OP_ATOM -> case stack of
    (h : t) -> case h of
      VAL_PAIR _ _ -> secd {stack = VAL_BOOL False : t, code = xs}
      _ -> secd {stack = VAL_BOOL True : t, code = xs}
    _ -> error "OP_ATOM expected one argument"
  OP_EQ -> case stack of
    (lhs : rhs : t) ->
      if (lhs == rhs)
        then secd {stack = VAL_BOOL True : t, code = xs}
        else secd {stack = VAL_BOOL False : t, code = xs}
    _ -> error "OP_EQ expected two arguments"
  OP_ADD -> case stack of
    (VAL_NUM x : VAL_NUM y : t) -> secd {stack = VAL_NUM (x + y) : t, code = xs}
    _ -> error "OP_ADD expected two integers"
  OP_MUL -> case stack of
    (VAL_NUM x : VAL_NUM y : t) -> secd {stack = VAL_NUM (x * y) : t, code = xs}
    _ -> error "OP_MUL expected two integers"
  _ -> secd {code = xs}

procAll :: SECD -> IO SECD
procAll Empty = return Empty
procAll secd@SECD {code = x : xs} =
  if x == OP_STOP
    then do return secd
    else procAll . proc $ secd

empty code =
  SECD
    { stack = [],
      env = [],
      code,
      dump = Empty
    }

var =
  let code =
        [ OP_LAB "x" [OP_LDC $ VAL_NUM 6],
          OP_LD "x",
          OP_STOP
        ]
   in empty code

cond =
  let code =
        [ OP_LDC $ VAL_BOOL True,
          OP_SEL [OP_LDC $ VAL_NUM 1] [OP_LDC $ VAL_NUM 2],
          OP_STOP
        ]
   in empty code

id' =
  let code =
        [ OP_LAB "id" [OP_LAM "x" [OP_LD "x", OP_RET]],
          OP_LDC $ VAL_BOOL True,
          OP_LD "id",
          OP_AP,
          OP_STOP
        ]
   in empty code

-- fact :: SECD
-- fact =
--     empty
--     [ OP_LAB
--         "fact"
--         [ OP_REC
--             "n"
--             [ OP_LD "n",
--                 OP_LDC (VAL_NUM 0),
--                 OP_EQ,
--                 OP_SEL
--                 [OP_LDC (VAL_NUM 1)]
--                 [ OP_LD "n",
--                     OP_LD "n",
--                     OP_LDC (VAL_NUM (-1)),
--                     OP_ADD,
--                     OP_LD "fact",
--                     OP_AP,
--                     OP_MUL
--                 ],
--                 OP_RET
--             ]
--         ],
--         OP_LDC (VAL_NUM 5),
--         OP_LD "fact",
--         OP_AP,
--         OP_STOP
--     ]

fact :: SECD
fact =
  empty
    [ OP_LAB
        "fact"
        [ OP_LD "n",
          OP_LDC (VAL_NUM 0),
          OP_EQ,
          OP_SEL
            [OP_LDC (VAL_NUM 1)]
            [ OP_LD "n",
              OP_LD "n",
              OP_LDC (VAL_NUM (-1)),
              OP_ADD,
              OP_LD "fact",
              OP_AP,
              OP_MUL
            ],
          OP_RET
        ],
      OP_LDC (VAL_NUM 5),
      OP_LD "fact",
      OP_AP,
      OP_STOP
    ]

cons =
  let code =
        [ OP_LDC VAL_NIL,
          OP_LDC $ VAL_NUM 1,
          OP_CONS,
          OP_LDC $ VAL_NUM 2,
          OP_CONS,
          OP_LDC $ VAL_NUM 3,
          OP_CONS,
          OP_STOP
        ]
   in empty code

add =
  let code =
        [ OP_LDC $ VAL_NUM 2,
          OP_LDC $ VAL_NUM 2,
          OP_ADD,
          OP_STOP
        ]
   in empty code

main :: IO SECD
main = procAll $ fact
